### PR TITLE
IR-195: set pod topology spread constraint

### PR DESCRIPTION
**Pull request by flavianmissi**
Defaults to:

```
  topologySpreadConstraints:
  - maxSkew: 1 
    topologyKey: topology.kubernetes.io/zone 
    whenUnsatisfiable: DoNotSchedule
    labelSelector:
      matchLabels:
        docker-registry: default
  - maxSkew: 1 
    topologyKey: kubernetes.io/hostname 
    whenUnsatisfiable: DoNotSchedule
    labelSelector:
      matchLabels:
        docker-registry: default
  - maxSkew: 1 
    labelSelector:
      matchLabels:
        docker-registry: default
    topologyKey: node-role.kubernetes.io/worker
    whenUnsatisfiable: DoNotSchedule
```

Users are able to override it via ImageRegistry Spec. In that case, only the user provided constraints will apply (the default constraints are removed).

~With the default constraints, it is _not_ possible to schedule pods on master nodes, regardless of tolerations or node selectors added.~
When node selectors are specified, no default constraints are used. We do this to maintain current behaviour where it's possible to deploy the image registry to specific nodes that may otherwise be excluded by the topology constraints.

Relates to [IR-195](https://issues.redhat.com/browse/IR-195)
Depends on https://github.com/openshift/cluster-image-registry-operator/pull/728 and https://github.com/openshift/api/pull/1052

~NOTE: this branch has `flavianmissi:upgrade-openshift-api` rebased into it. Please ignore the vendored changes, they should be removed when https://github.com/openshift/cluster-image-registry-operator/pull/728 lands and we update openshift/api to a commit including https://github.com/openshift/api/pull/1052~

### CODE REVIEW

**Comment by dmage** [comment]
/lgtm

/hold
until `replace` is removed after approvals

/assign @xiuwang @bmcelvee @sferich888 
for approvals

**Comment by dmage** [comment]
/assign @bmcelvee @sferich888 
for Docs and Px approvals

File: pkg/resource/podtemplatespec.go [hunk 1]
```diff
@@ -378,9 +378,14 @@ func makePodTemplateSpec(coreClient coreset.CoreV1Interface, proxyLister configl
 		resources = *cr.Spec.Resources
 	}
 
+	nodes, err := coreClient.Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: "topology.kubernetes.io/zone"})
```
Comment (dmage) [inline pkg/resource/podtemplatespec.go:381]: err is not checked

Comment (flavianmissi) [reply pkg/resource/podtemplatespec.go:381]: fixed

File: pkg/resource/podtemplatespec.go [hunk 2]
```diff
@@ -378,45 +378,83 @@ func makePodTemplateSpec(coreClient coreset.CoreV1Interface, proxyLister configl
...
+	if err != nil {
+		return corev1.PodTemplateSpec{}, deps, fmt.Errorf("Could not check nodes for zone failure domain: %s", err)
```
Comment (dmage) [inline pkg/resource/podtemplatespec.go:383]:
```suggestion
		return corev1.PodTemplateSpec{}, deps, fmt.Errorf("could not check nodes for zone failure domain: %s", err)
```

**Comment by KeenonLee** [comment]
Test again, nodeSelector and tolerations works well, thanks
/label qe-approved

**Comment by dmage** [comment]
/label px-approved
IR-228 isn't marked as TE Needed

**Pull request merged**